text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,723 @@
+
+The type of the expression `become <call>` is `!` (the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html)). This is consistent with other control flow constructs such as `return`, which also have the type of `!`.
+
+Note that as `become` is a keyword reserved for exactly the use-case described in this RFC there is no backward-compatibility break. This RFC only specifies the use of `become` inside of functions and instead leaves usage outside of functions unspecfied for use by other features.
+
+This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.
+
+See below for specifics on interations with other features.
+
+## Coercions of the Tail Called Function's Return Type
+
+All coercions that do any work (like deref coercion, unsize coercion, etc) are prohibited.
+Lifetime-shortening coercions (`&'static T` -> `&'a T`) are allowed but will be checked by the borrow checker.
+
+Note that, while in theory, never-to-any coercions (`! -> T`) could be allowed, they are difficult to implement and require backend support. As a result they are not allowed as per this RFC. This has no effect on using macros like `panic!()` as they are not functions, affected are only functions like the following:
+
+```rust
+fn never() -> ! {
+    loop {}
+}
+
+fn tail_call_never_type() -> usize {
+    become never();
+}
+```
+
+## Closures
+[closures]: #closures
+
+Tail calling closures _and_ tail calling _from_ closures is **not** allowed.
+This is due to the high implementation effort, see below, this restriction can be lifted by a future RFC.
+
+Closures use the `rust-call` unstable calling convention, which would need to be adapted to guarantee TCE.
+Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.
+
+## Variadic functions using `c_variadic`
+
+Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.
+As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.
+
+## Generators
+
+Tail calling [generators](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) is **not** allowed as it is a fundamental mismatch of functionality. Generators expect to be called multiple times `yield`ing values, however, when using a tail call control would never be returned to the calling function.
```

Comment:
Not sure what is meant by this, you can't call a generator anyway (and tail-calling `resume` should be possible, it's just a method). I'd only leave the "from generators" restriction.

Reply:
updated: https://github.com/rust-lang/rfcs/pull/3407/commits/4b184876a5334f2e6f4b445e7cc590ee4cbf2137